Document some of the basic cmake stuff that is unique to us (#244)
NUbots/NUbook: src/book/02-system/01-foundations/02-build-system.mdx

    "asm_object": "elf64",
}
```

### Dependency Declaration

CMake needs to know what dependencies we have and where to find them.
In a module's folder the `CMakeLists.txt` file tells CMake what is needed to build a module.

> The `nuclear_modlue()` function tells CMake that this directory contains a module that may be added to a role.

Reviewer: ```suggestion
The `nuclear_module()` function tells CMake that this directory contains a module that may be added to a role.
```

This function may take named parameters.

| Parameter  | Usage                                                                          |
| ---------- | ------------------------------------------------------------------------------ |

> | LANGUAGE   | The Programming Language that this module is in (currently only c++)           |

Reviewer: ```suggestion
| LANGUAGE   | The programming language that this module is in (currently only C++)           |
```

| LIBRARIES  | The libraries that this module depends on and must be linked with              |
| SOURCES    | Extra code outside the `src` directory that must be built                      |

> | DATA_FILES | Extra data files outside the `data` directory that is necessary for the module |

Reviewer: ```suggestion
| DATA_FILES | Extra data files outside the `data` directory that the module depends on |
```
Not sure if this is right but the sentence felt incomplete/grammar dodgy

Author: ```suggestion
| DATA_FILES | Extra data files outside the `data` directory that the module uses |
```

> Now for libraries in particular CMake needs to know where the libraries are located,
> the function `find_package()` is used to tell CMake to find the library.
> 
> CMake then looks in the `cmake/Modules/` directory for files like `FindX.cmake` for the logic to find the files.
> CMake can also look for build instruction files in `lib/cmake/X/` like `XConfig.cmake` where `X` is the name of the
> library.

Reviewer: ```suggestion
CMake needs to know where libraries are located. The `cmake/Modules/` directory contains files called `FindX.cmake`, where `X` is the name of a library. These files contain the logic for finding library files. The function  `find_package(X)` is used to tell CMake to look for one of these `.cmake` files and run it.

CMake can also look for build instruction files in `/lib/cmake/X/XConfig.cmake` where `X` is the name of the library.
```

> An example file looks like

Reviewer: ```suggestion
An example of one of these files, `FindAravis.cmake` which finds the Aravis library, is shown below
```

| HEADER                   | Any header, e.g. `.h` or `.hpp`, files that define the api of the library                     |
| LIBRARY                  | The name of the library on disk, e.g. if we have `/usr/lib/libaravis-0.8.so` it'd be `aravis` |
| LIBRARIES                | A list instead of a single LIBRARY                                                            |
| PATH_SUFFIX              | Add subdirectories to look for includes and libraries in the normal search directories        |
| BINARY                   | A binary that can be used to extract the version number                                       |
| VERSION_BINARY_ARGUMENTS | The arguments to pass to the binary that will result in a version number being printed        |
| VERSION_FILE             | A file that contains information on the version number of the library                         |
| VERSION_REGEX            | The regex used on the version file to extract the version number                              |
| LINK_TYPE                | The type of linkage that the library will use: `UNKNOWN` is the default and will pick either, |

> |                          | `SHARED` will use dynamic linkage and look for a `.so` library file                           |
> |                          | `STATIC` will use static linkage and look for a `.a` library file                             |

Reviewer: ```suggestion
|      SHARED       | No arguments; use dynamic linkage and look for a `.so` library file                           |
|        STATIC     | No arguments; use static linkage and look for a `.a` library file                             |
```
maybe something like this?

Reviewer: will need formatting tho

Author: `Unknown`, `Shared` and `Static` are options to link type.

Reviewer: Oh I see, in that case put it on the one line? It looks like boolean flags here, I assumed each row of the table was a separate parameter.
